frontend/packages/console-app/src/components/cloud-shell/CloudShellTerminal.tsx

```diff
@@ -32,37 +36,81 @@ const resource = {
 };
 
 const CloudShellTerminal: React.FC<CloudShellTerminalProps> = ({ username, onCancel }) => {
-  const [data, loaded, loadError] = useK8sWatchResource<CloudShellResource>(resource);
+  const [data, loaded, loadError] = useK8sWatchResource<CloudShellResource[]>(resource);
+  const [initData, setInitData] = React.useState<TerminalInitData>();
+  const [initDataLoading, setInitDataLoading] = React.useState<boolean>(false);
+  const [initError, setInitError] = React.useState<string>();
+  const [workspaceNamespace, setWorkspaceNamespace] = React.useState<string>();
 
-  if (loadError) {
+  React.useEffect(() => {
+    let destroy = false;
+    if (Array.isArray(data)) {
+      const workspace = data.find(
+        (ws) => ws?.metadata?.annotations?.[CLOUD_SHELL_USER_ANNOTATION] === username,
+      );
+      const running = workspace?.status?.phase === 'Running';
+
+      if (running) {
+        setInitDataLoading(true);
+        const { name, namespace } = workspace.metadata;
+        initTerminal(username, name, namespace)
+          .then((res: TerminalInitData) => {
+            if (destroy) return;
+            setInitData(res);
+            setInitDataLoading(false);
+            setWorkspaceNamespace(namespace);
+          })
+          .catch(() => {
+            if (destroy) return;
+            setInitDataLoading(false);
+            setInitError('Failed to connect to your OpenShift command line terminal');
+          });
+      }
+    }
+
+    return () => {
+      destroy = true;
+    };
+  }, [data, username]);
```

Comment:
We probably only need to enter this effect if `username` or workspace phase or namespace changes.